Support exact decimal value to preserve precision

Akanksha-kedia opened the pull request:
### Core Fix in RequestUtils.java

__File__: `pinot-common/src/main/java/org/apache/pinot/common/utils/request/RequestUtils.java` __Line 238__: Replaced the precision-losing code with:

```java
// Store exact decimal value to preserve precision
literal.setBigDecimalValue(BigDecimalUtils.serialize(bigDecimalValue));
```

Jackie-Jiang commented:
This problem is more complicated than just changing the literal representation. We need to ensure server side can handle big decimal properly, then change the broker side literal representation in the next release

Akanksha-kedia committed:
Support exact decimal value to preserve precision

Phase 1 of a two-phase fix for decimal literal precision loss in queries.

Problem: SQL decimal literals (e.g. WHERE amount > 1.23456789) are silently
downcast to double in RequestUtils.getLiteral(SqlLiteral), losing precision
for values that cannot be represented exactly as IEEE 754 doubles.

This PR (Phase 1):
- Documents the backward-compatibility constraint: the broker must continue
  sending DOUBLE_VALUE until all servers handle BIG_DECIMAL_VALUE
- Adds a Javadoc to getLiteral(SqlLiteral) explaining the two-phase plan
- Adds tests proving the server (LiteralContext) already handles
  BIG_DECIMAL_VALUE correctly, so Phase 2 can proceed without a
  coordinated cluster upgrade

Phase 2 (follow-up): change getLiteral(SqlLiteral) to setBigDecimalValue()
once all servers are upgraded.

Akanksha-kedia commented:
@Jackie-Jiang thank you for the guidance. I have reworked the PR to follow the two-phase approach you described.

**What this PR (Phase 1) does:**
- Keeps `setDoubleValue()` in `getLiteral(SqlLiteral)` for backward compatibility — no wire-format change yet
- Tests in `LiteralContextTest` prove the server already deserializes `BIG_DECIMAL_VALUE` with full precision
- Tests in `RequestUtilsTest` document the current DOUBLE_VALUE encoding contract
- Javadoc on `getLiteral(SqlLiteral)` explains when and why to switch to `setBigDecimalValue()`

**Phase 2 (follow-up PR):** switch `setDoubleValue()` to `setBigDecimalValue()` once all servers are upgraded.

All CI checks pass. Could you please review and approve so Phase 1 can be merged?

Akanksha-kedia commented:
@xiangfu0

Akanksha-kedia commented:
@xiangfu0 please review

Jackie-Jiang reviewed (commented):
Seems there is no change to production code?

Akanksha-kedia commented:
@Jackie-Jiang — thank you for the guidance on the two-phase approach. I've restructured the PR per Phase 1 (keep `setDoubleValue()` for backward compatibility, document the contract, add tests proving server already deserializes `BIG_DECIMAL_VALUE` correctly). Phase 2 (switch to `setBigDecimalValue()`) will be a follow-up once servers are rolled out.

Could you please take another look at the updated PR and let me know if the Phase 1 approach is aligned with what you had in mind?

Akanksha-kedia commented:
Hi @Jackie-Jiang — pinging for a re-review. All CI checks are green. Please let me know if there are any further changes needed!

Akanksha-kedia commented:
cc @walterddr @61yao @Jackie-Jiang — would appreciate a re-review when you get a chance. Happy to address any remaining concerns.
